pkg/complete/line.go

```diff
@@ -0,0 +1,101 @@
+
+package complete
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+)
+
+// LineReader has three things and returns on string.
+// The three things are an io.Reader, an io.Writer, and a Completer
+// Bytes are read one at a time, and depending on their value,
+// are written to the io.Writer.
+// If the completer returns 0 or 1 answers, a 0 or 1 length string is returned.
+// If there are two or more answers, they are printed and the line is
+// printed out again.
+// Most characters are just echoed.
+// Special handling: newline or space returns.
+// tab tries to complete
+// backspace erases. Since everything is ansi, we assume ansi.
+//
+// LineReader is used to implement input for a Completer.
+// It uses a Completer, io.Reader, io.Writer, and bytes.Buffer.
+// bytes are read from the reader, processed, held in the
+// bytes.Buffer and, as a side effect, some information is
+// written to the io.Writer.
+type LineReader struct {
+	// Completer for this LineReader
+	C Completer
+	// R is used for input. Most characters are stored in the
+	// Line, while some initiate special processing.
+	R io.Reader
+	// W is used for output, usually for showing completions.
+	W io.Writer
+	// Lines holds incoming data as it is read.
+	Line bytes.Buffer
+}
+
+// NewLineReader returns a LineReader.
+func NewLineReader(c Completer, r io.Reader, w io.Writer) *LineReader {
+	return &LineReader{C: c, R: r, W: w}
+}
+
+// ReadOne tries to read one choice from l.R, printing out progress to l.W.
+// In the case of \t it will try to complete given what it has.
+// It will show the result of trying to complete it.
+// If there is only one possible completion, it will return the result.
+// In the case of \r or \n, if there is more than one choice,
+// it will return the list of choices, preprended with that has been typed so far.
+func (l *LineReader) ReadOne() ([]string, error) {
+	for {
+		var b [1]byte
+		n, err := l.R.Read(b[:])
+		if err != nil {
+			if err == io.EOF {
+				ln := l.Line.String()
+				return l.C.Complete(ln[:len(ln)-1])
+			}
+			return nil, err
+		}
+		Debug("ReadOne: got %v, %v, %v", b[0], n, err)
+		if n == 0 {
+			continue
+		}
+		switch b[0] {
+		default:
+			Debug("Just add it to line and pipe")
+			l.Line.Write(b[:])
+			l.W.Write(b[:])
+		case '\n', '\r':
+			ln := l.Line.String()
+			s, err := l.C.Complete(ln[:len(ln)-1])
+			Debug("Newline, just return %v", s)
+			// the choice to use is always the first element.
+			// In the case too many elements, put what they
+			// typed so far as the only choice
+			if len(s) > 1 {
+				s = append([]string{ln}, s...)
+			}
+			return s, err
+		case '\t':
+			Debug("Try complete with %s", l.Line.String())
+			s, err := l.C.Complete(l.Line.String())
+			Debug("Complete returns %v, %v", s, err)
+			if err != nil {
+				return nil, err
+			}
+			if len(s) < 2 {
+				Debug("Return is %v", s)
+			}
+			if _, err := l.W.Write([]byte(fmt.Sprintf("\n\r%v\n\r%v", s, l.Line.String()))); err != nil {
```

Comment:
Is there a reason to use `\n\r` instead of `\r\n`? I see `\r\n` used a lot more often.

Reply:
ryan, fantastic review, I will respond more letter but I had to respond to this one.
back in the day, CR-LF made lots of sense. Typically it took the print head on an ASR-33 teletype longer to go back to the left than to roll the paper up a line. So CR-LF made lots of sense. 

but consider a text display. What takes longer, scroll up a line or start drawing at the left? At some point I saw it evolve to NL-CR.

Nowadays, which sequence is faster? Both? Neither? I doubt it even matters. But it's more likely the cast that \n\r is better than \r\n than otherwise. 

Anyway, thanks much for the review, I'll fix things. And, yes, we want to support unicode, maybe.